_posts/2020-07-12-ray-tune.md

```diff
@@ -0,0 +1,264 @@
+> stopping, distributed execution) can provide significant speedups over grid
+> search and random search.
+>
+> However, the machine learning ecosystem is missing a solution that provides
+> users with the ability to leverage these new algorithms while allowing users
+> to stay within the Scikit-Learn API. In this blog post, we introduce
+> tune-sklearn to bridge this gap. Tune-sklearn is a drop-in replacement for
+> Scikit-Learn’s model selection module with state-of-the-art optimization
+> features.
+>
+> —[GridSearchCV 2.0 — New and Improved][1]
+
+This exaggerates Ray's novelty because Dask-ML and auto-sklearn also provide
+"users with the ability to leverage these new algorithms while allowing users
+to stay within the Scikit-Learn API." Ray is not the first project to "bridge
+this gap."
+
+That is, Dask-ML offers the same features that Ray's tune-sklearn offers:
+
+[dml-hod]:https://ml.dask.org/hyper-parameter-search.html
+
+> Here’s what tune-sklearn has to offer:
+> 1. **Consistency with Scikit-Learn API** ...
+> 2. **Modern hyperparameter tuning techniques** ...
+> 3. **Framework support** ...
+> 4. **Scale up** ...
+>
+> Tune-sklearn is also **fast**.
+
+Dask-ML has had these features since June 2019, about a year earlier than Ray.
+Let's look at each feature and see what Dask-ML and Ray have to offer:
+
+> 1\. **Consistency with Scikit-Learn API**: tune-sklearn is a drop-in
+> replacement for GridSearchCV and RandomizedSearchCV, so you only need to
+> change less than 5 lines in a standard Scikit-Learn script to use the API.
+
+Dask-ML is also consistent with the Scikit-learn API. Here's how to use
+Dask-ML's hyperparameter optimization:
+
+``` python
+from dask_ml.model_selection import HyperbandSearchCV
+from scipy.stats import loguniform
+from sklearn.datasets import make_classification
+from sklearn.linear_model import SGDClassifier
+
+X, y = make_classification()
+model = SGDClassifier()
+params = {"loss": ["hinge", "log", "modified_huber", "squared_hinge", "perceptron"],
+          "alpha": loguniform(1e-5, 1e-3)}
+
+search = HyperbandSearchCV(model, params)
+search.fit(X, y, classes=[0, 1])
+```
+
+This example is almost identical to [one of Ray's examples][sgd-eg] and [one of
+Scikit-learn's examples][skl-eg]. Clearly, Dask-ML offers a Scikit-learn
+compatible model selection API.
+
+[skl-eg]:https://scikit-learn.org/stable/auto_examples/model_selection/plot_randomized_search.html#sphx-glr-auto-examples-model-selection-plot-randomized-search-py
+[sgd-eg]:https://github.com/ray-project/tune-sklearn/blob/31f228e21ef632a89a74947252d8ad5323cbd043/examples/sgd.py
+
+> 2\. **Modern hyperparameter tuning techniques:** tune-sklearn is the only
+> Scikit-Learn interface that allows you to easily leverage Bayesian
+> Optimization, HyperBand, and other optimization techniques by simply toggling
+> a few parameters.
+
+Dask-ML also offers modern hyperparameter tuning techniques. Hyperband is the
```

Comment:
Injecting a personal option: whether "modern = good" is subjective :) Sticking to objective truths, we can say that Dask-ML implements Hyperband and (currently) lacks a Bayesian Optimization method.

I'd also note that that we've built a pretty flexible framework for building *custom* asynchronous training strategies.

Reply:
Thanks! That suggestion has generated an intuitive explanation of how Hyperband behaves.

Reply:
> pretty flexible framework for building custom asynchronous training strategies.

Thanks! That's a good point.